tools/print_tock_memory_usage.py

```diff
@@ -0,0 +1,422 @@
+            return ("  " + key + str(group_size) + " bytes\n")
+
+def print_groups(title, groups):
+    """Print title, then all of the variable groups in groups."""
+    group_sum = 0
+    output = ""
+    max_string_len = len(max(groups.keys(), key=len))
+    for key in sorted(groups.keys()):
+        symbols = groups[key]
+
+        group_size = 0
+        for (_, size) in symbols:
+            group_size = group_size + size
+
+        output = output + string_for_group(key, max_string_len, group_size, len(symbols))
+        group_sum = group_sum + group_size
+
+    print(title + ": " + str(group_sum) + " bytes")
+    print(output, end = ' ')
+
+def print_symbol_information():
+    """Print out all of the variable and function groups with their flash/RAM
+       use."""
+    variable_groups = {}
+    group_symbols(variable_groups, kernel_initialized, show_waste, "RAM")
+    group_symbols(variable_groups, kernel_uninitialized, show_waste, "Flash+RAM")
+    if (show_waste):
+        print() # Place an newline after waste reports
+
+    print_groups("Variable groups (RAM)", variable_groups)
+
+    print()
+    print("Embedded data (in flash): " + str(padding_text) + " bytes")
+    print()
+    function_groups = {}
+    # Embedded constants in code (e.g., after functions) aren't counted
+    # in the symbol's size, so detecting waste in code has too many false
+    # positives.
+    group_symbols(function_groups, kernel_functions, False, "Flash")
+    print_groups("Function groups (in flash)", function_groups)
+    print()
+
+def get_addr(symbol_entry):
+    """Helper function for sorting symbols by start address."""
+    return symbol_entry[1]
+
+def compute_padding(symbols):
+    """Calculate how much padding is in a list of symbols by comparing their
+       reporting size with the spacing with the next function and return
+       the total differences."""
+    symbols.sort(key=get_addr)
+    func_count = len(symbols)
+    diff = 0
+    for i in range(1, func_count):
+        (esymbol, eaddr, esize, _) = symbols[i - 1]
+        (_, laddr, _, _) = symbols[i]
+        total_size = laddr - eaddr
+        symbols[i - 1] = (esymbol, eaddr, esize, total_size)
+        if total_size != esize:
+            diff = diff + (total_size - esize)
+
+    return diff
+
+def parse_options(opts):
+    """Parse command line options."""
+    global symbol_depth, verbose, show_waste
+    valid = 'd:vs'
+    long_valid = ['depth=', 'verbose', 'show-waste']
+    optlist, leftover = getopt.getopt(opts, valid, long_valid)
+    for (opt, val) in optlist:
+        if opt == '-d' or opt == '--depth':
+            symbol_depth = int(val)
+        elif opt == '-v' or opt == '--verbose':
+            verbose = True
+        elif opt == '-s' or opt == '--show-waste':
+            show_waste = True
+        else:
+            usage("unrecognized option: " + opt)
+            return []
+
+    return leftover
+
+ # Script starts here ######################################
```

Comment:
I would prefer if everything from here on was moved into a `if __name__ == "__main__":` block, so that this module can be imported by another Python file without running the script itself (I am looking into what it will take to call this file from Travis and it will likely involve calling it from another python script).

This would also require niklas suggestion of moving away from global variables for `symbol_depth`, `verbose`, and `show_waste` and instead passing those to each function that uses them.

Reply:
Will do.

Reply:
I actually re-architected my Travis script to call your script directly, so up to you whether to still make this change

Reply:
Done. Note the script now invokes llvm-objdump.